### DISCUSSION TITLE
[resolved] Why don't .NET use CAS in ConcurrentDictionary Get?

### DISCUSSION THREAD

**Discussion by Cyl18** [dotnet/runtime#48187] [Q&A]
[resolved]

> https://github.com/dotnet/runtime/blob/cb5431d0c0790a69925145a15952a9601b9be102/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L486
> `Volatile.Read` inserts a memory barrier which might be performance consuming.
> Why not to use CAS? And what's the difference between CAS and memory barrier?

**Comment by LeaFrock** [comment]
So... why is this question resolved?

**Reply by huoyaoyuan** [reply]
The question is misunderstanding about "memory barrier".

The barrier means **correctness of execution order**. It has no relationship with "security barrier".

**Reply by PathogenDavid** [reply]
@huoyaoyuan Pretty sure CAS in this context is compare-and-swap (IE: `Interlocked.CompareExchange`), not code access security. (The latter is what my mind jumped to as well though.)
